### ISSUE TITLE
Sudden error - Refused to execute script from worker.ts because it's MIME type (video/mp2t) is not executable

### ISSUE THREAD

**Issue by polisen** [vercel/next.js#31009]
### What version of Next.js are you using?

12.0.1

### What version of Node.js are you using?

14

### What browser are you using?

chrome

### What operating system are you using?

macOS

### How are you deploying your application?

next start

### Describe the Bug

I was writing code - plugging around with webworkers. It was literally working perfectly - worker running, hit save, and now Next.js considers the file a (video/mp2t)...wat?

I L.I.T.E.R.A.L.L.Y just hit save. No server reboot, config change no nothing. It was working.

Here's the worker if anyone smarter than me can identify some magic phrase or something idk.

I really am stuck here, please help.

### Expected Behavior

for it to load in a correct manner.

### To Reproduce

Idk how you would go about reproducing this.

Here's my repo - the branch firebase-integration is required.

https://github.com/polisen/browser-file-bundler/tree/firebase-integration.

Re-installing the whole repo did nothing.

**Comment by polisen** [comment]
![Screenshot 2021-11-05 at 08 27 29](https://user-images.githubusercontent.com/29376308/140477829-b2728d63-57a1-48e7-bafd-23f38ac53634.png)

**Comment by GN-c** [comment]
I've had same issue, looks like it's caused by webpack.
since webpack v5, I think it's detecting worker scripts by matching following pattern:
`new Worker(new URL('path', import.meta.url));`
even defining worker like this does not work:
![image](https://user-images.githubusercontent.com/64430939/171948733-45e48e31-d6ea-4221-8971-f422794f0798.png)
or by defining class which extends worker:
![image](https://user-images.githubusercontent.com/64430939/171948836-f1e3e79a-43ac-445b-a682-e47544f033bf.png)

currently it's not causing much of a problem, but it's better to see this fixed

**Comment by theahura** [comment]
has anyone already opened a bug for this on webpack?

**Comment by christoph-pflueger** [comment]
I can confirm this bug.

```javascript
const url = new URL('../workers/worker.ts', import.meta.url)
const worker = new Worker(url)
```

... results in the aforementioned MIME type error.

```javascript
const url = '../workers/worker.ts'
const worker = new Worker(new URL(url, import.meta.url))
```

... results in `DOMException: Failed to construct 'Worker': Script at 'file:///home/dev/projects/test/src/workers/worker.ts' cannot be accessed from origin 'http://localhost:3000'.`

```javascript
const worker = new Worker(new URL('../workers/worker.ts', import.meta.url))
```

... works just fine.

**Comment by hiromaily** [comment]
Is there any progress? I'm still facing this error.

**Comment by maxnowack** [comment]
@hiromaily the workaround is to construct the worker object with everything in the same line. if you don't use variables and don't use different lines, it should be working.
> https://github.com/vercel/next.js/issues/31009#issuecomment-1338645354
> ```js
> const worker = new Worker(new URL('../workers/worker.ts', import.meta.url))
> ```
> 
> ... works just fine.

**Comment by hiromaily** [comment]
@maxnowack 
Thanks for your advise. Actually I tried its pattern. My code is like the following
```
export class MyWorker extends Worker {
  constructor() {
    super(new URL("./myWorker.ts", import.meta.url));
  }
```

**Comment by maxnowack** [comment]
@hiromaily right, extending the class doesn't work. You really have to instantiate everything in one line.

**Comment by adamczykm** [comment]
Confirming the bug and the solution. 
This kind of bugs make me question the choice of my profession.

**Comment by mulgurul** [comment]
Anyone knows if there is a way to turn of webpack interference and manage webworker load yourself

**Comment by n1ru4l** [comment]
I noticed this happening when using `next dev --turbo`.  As mentioned in [this blog post](https://loclv.hashnode.dev/a-simple-web-worker-demo-with-typescript-and-nextjs), removing the `--turbo` flag solves the issue for me.

**Comment by WebReflection** [comment]
so ... people are filing issues [in my projects](https://github.com/WebReflection/coincident/issues/44) because something here doesn't understand *standard ECMAScript*.

I am particularly keen to help this comment from 2022:
https://github.com/vercel/next.js/issues/31009#issuecomment-1324330092

I don't know the current state though ... and I don't understand why removing `--turbo` would work, if it does ... can anyone please summarize to me why a developer should not be able to write standard *JS* or extend a class that is possible since 2015?

Thank you, I mean it!

**Comment by paschaldev** [comment]
@WebReflection 

Spent hours trying to debug this problem also

JavaScript... The gift/curse that keeps on giving

**Comment by WebReflection** [comment]
@paschaldev this bug has nothing to do with JS itself, it's the `--turbo` flag that affects some tool incapable of dealing with standard ECMAScript but of course we blame JS and not the tool 😔

**Comment by gndelia** [comment]
I am seeing this issue without using the `--turbo` flag (using next 14, which still uses `webpack` I believe)

**Comment by ApacheEx** [comment]
I can confirm it works with nextjs v15 and `--turbopack`

the main issue is how you initialize worker, it only works with the following approach:

```
const worker = new Worker(new URL('../workers/my-worker', import.meta.url))
```

**Comment by LmingXie** [comment]
> I can confirm it works with nextjs v15 and `--turbopack`
> 
> the main issue is how you initialize worker, it only works with the following approach:
> 
> ```
> const worker = new Worker(new URL('../workers/my-worker', import.meta.url))
> ```

The problem still exists after building. Is it because of packaging through Webpack?

**Comment by WebReflection** [comment]
@LmingXie I assume so, `import.meta.url` cannot resolve locally what the real value would be once deployed ... unless there is a config for that, as in: "*use this fully qualified URL to resolve relative paths*"

**Comment by LmingXie** [comment]
new URL() supports environment variables, so you can pass NEXT_PUBLIC_WEB_WORKER_PATH=/_next/static/chunks/workers.js, but there is still a packaging problem. I compile and package in the following way:
```
      config.module.rules.push({
        test: /\.worker\.ts$/,
        include: [path.resolve(process.cwd(), 'workers')],
        use: [
          {
            loader: 'babel-loader', 
          },
          {
            loader: 'worker-loader',
            options: {
              filename: 'static/chunks/workers/[name].js',
              chunkFilename: 'static/chunks/workers/[name].js', 
              publicPath: '/_next/',
            },
          },
        ],
      });
```
Loading via new URL() will automatically compile .ts files, but nextJS 15 and webpack have issues with supporting file packaging for new URL(), and no files are generated after packaging.
Therefore, the packaging obfuscation logic of Worker files is handled separately, and loading is performed based on different environments.

```
            const worker = asProd ? new Worker(`${process.env.NEXT_PUBLIC_WEB_WORKER_PATH}/worker.js`)
                : new Worker(new URL('/workers/worker', import.meta.url));
```

**Comment by marcoshernanz** [comment]
Closing this as a documented Webpack limitation rather than a Next.js MIME handling bug. I reproduced it on Next.js 12.0.1 and 16.3.2 with Webpack: when the `URL` is stored in a variable, Webpack emits the `.ts` source as a regular asset, so it is served as `video/mp2t` and the browser rejects it.

Keep the worker construction in the statically analyzable form documented by Webpack:

```js
new Worker(new URL('./worker.ts', import.meta.url))
```

That form works with Webpack, and the hoisted form works with the default Turbopack bundler in current Next.js. Related worker-creation cases were also closed in #61787 and #72376.
